File: 2024/day6/part1.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"time"
)

const (
	UP = iota
	RIGHT
	DOWN
	LEFT
)

type Position struct {
	row int
	col int
}

type Guard struct {
	dir int
	pos Position
}

func (g Guard) NextPosition() Position {
	switch g.dir {
	case UP:
		return Position{g.pos.row - 1, g.pos.col}
	case RIGHT:
		return Position{g.pos.row, g.pos.col + 1}
	case DOWN:
		return Position{g.pos.row + 1, g.pos.col}
	case LEFT:
		return Position{g.pos.row, g.pos.col - 1}
	default:
		panic("invalid dir: " + string(g.dir))
	}
}

func (g *Guard) Turn() {
	g.dir = (g.dir + 1) % 4
}

func (g *Guard) Move(pos Position) {
	g.pos = pos
}

func (g Guard) InBounds(numRows int, numCols int) bool {
	return 0 <= g.pos.row &&
		g.pos.row < numRows &&
		0 <= g.pos.col &&
		g.pos.col < numCols
}

func read(filename string) (map[Position]bool, Guard, int, int) {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	obstacles := make(map[Position]bool)

	row := 0
	scanner := bufio.NewScanner(file)
	var guard Guard
	var numCols int
	for scanner.Scan() {
		line := scanner.Text()
		numCols = len(line)
		for col, c := range line {
			if string(c) == "#" {
				obstacles[Position{row, col}] = true
			} else if string(c) == "^" {
				guard = Guard{UP, Position{row, col}}
			}
		}
		row += 1
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return obstacles, guard, row, numCols
}

func run() int {
	obstacles, guard, numRows, numCols := read(os.Args[1])
	visited := make(map[Position]bool)
	for guard.InBounds(numRows, numCols) {
		visited[guard.pos] = true
		nextPos := guard.NextPosition()
		if _, ok := obstacles[nextPos]; ok {
			guard.Turn()
		} else {
			guard.Move(nextPos)
		}
	}
	return len(visited)
}

func main() {
	startTime := time.Now()
	fmt.Println("RESULT:", run())
	elapsed := time.Since(startTime)
	fmt.Println("completed in", elapsed.Seconds(), "seconds")
}

File: 2024/day6/part1.py
import re
import time
from argparse import ArgumentParser
from enum import Enum
from dataclasses import dataclass

class Dir(bytes, Enum):
  def __new__(cls, value, char, row_change, col_change):
    obj = bytes.__new__(cls, [value])
    obj._value_ = value
    obj.char = char
    obj.row_change = row_change
    obj.col_change = col_change
    return obj
  def __repr__(self):
    return self.char
  UP = (0, "^", -1, 0)
  RIGHT = (1, ">", 0, 1)
  DOWN = (2, "v", 1, 0)
  LEFT = (3, "<", 0, -1)


@dataclass
class Guard:
  dir: Dir
  row: int
  col: int

  def turn(self):
    self.dir = Dir((self.dir.value + 1) % len(Dir))

  def next_pos(self):
    return self.row + self.dir.row_change, self.col + self.dir.col_change


def read(file):
  obstacles = set()
  guard = None
  with open(file, "r") as f:
    row = 0
    for line in f:
      for col, c in enumerate(line.strip()):
        if c == "#":
          obstacles.add((row, col))
        elif c == "^":
          guard = Guard(Dir.UP, row, col)
      row += 1
  return obstacles, guard, row, col + 1


def print_lab(obstacles, guard, positions, num_rows, num_cols):
  for row in range(num_rows):
    for col in range(num_cols):
      if (row, col) in obstacles:
        print("#", end="")
      elif guard.row == row and guard.col == col:
        print(guard.dir.char, end="")
      elif (row, col) in positions:
        print("X", end="")
      else:
        print(".", end="")
    print()


def main(file):
  obstacles, guard, num_rows, num_cols = read(file)
  positions = set()
  while 0 <= guard.row < num_rows and 0 <= guard.col < num_cols:
    positions.add((guard.row, guard.col))
    next_position = guard.next_pos()
    if next_position in obstacles:
      guard.turn()
    else:
      guard.row, guard.col = next_position
  return len(positions)


if __name__ == "__main__":
  parser = ArgumentParser()
  parser.add_argument("file")
  args = parser.parse_args()
  start_time = time.time()
  print("\nRESULT:", main(args.file))
  print("--- COMPLETED IN %s SECONDS ---" % (time.time() - start_time))